core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java

```diff
@@ -19,23 +19,665 @@
+                equalToJSon("{'string':[]}"));
+        assertResult(() -> builder().startObject().field("string").value((String) null).endObject(),
+                equalToJSon("{'string':null}"));
+        assertResult(() -> builder().startObject().field("string").value("42").endObject(),
+                equalToJSon("{'string':'42'}"));
+    }
+
+    public void testBinaryField() throws Exception {
+        assertResult(() -> builder().startObject().field("binary", (byte[]) null).endObject(),
+                equalToJSon("{'binary':null}"));
+
+        final byte[] randomBytes = randomBytes();
+        BytesReference bytes = builder().startObject().field("binary", randomBytes).endObject().bytes();
+
+        XContentParser parser = xcontentType().xContent().createParser(bytes);
+        assertThat(parser.nextToken(), is(Token.START_OBJECT));
+        assertThat(parser.nextToken(), is(Token.FIELD_NAME));
+        assertThat(parser.currentName(), equalTo("binary"));
+        assertThat(parser.nextToken().isValue(), is(true));
+        assertArrayEquals(randomBytes, parser.binaryValue());
+        assertThat(parser.nextToken(), is(Token.END_OBJECT));
+        assertNull(parser.nextToken());
+    }
+
+    public void testBinaryValue() throws Exception {
+        assertResult(() -> builder().startObject().field("binary").value((byte[]) null).endObject(),
+                equalToJSon("{'binary':null}"));
+
+        final byte[] randomBytes = randomBytes();
+        BytesReference bytes = builder().startObject().field("binary").value(randomBytes).endObject().bytes();
+
+        XContentParser parser = xcontentType().xContent().createParser(bytes);
+        assertThat(parser.nextToken(), is(Token.START_OBJECT));
+        assertThat(parser.nextToken(), is(Token.FIELD_NAME));
+        assertThat(parser.currentName(), equalTo("binary"));
+        assertThat(parser.nextToken().isValue(), is(true));
+        assertArrayEquals(randomBytes, parser.binaryValue());
+        assertThat(parser.nextToken(), is(Token.END_OBJECT));
+        assertNull(parser.nextToken());
+    }
+
+    public void testBinaryValueWithOffsetLength() throws Exception {
+        assertResult(() -> builder().startObject().field("binary").value(null, 0, 0).endObject(),
+                equalToJSon("{'binary':null}"));
+
+        final byte[] randomBytes = randomBytes();
+        final int offset = randomIntBetween(0, randomBytes.length);
+        final int length = randomIntBetween(1, Math.max(1, randomBytes.length - offset));
+
+        XContentBuilder builder = builder().startObject();
+        if (randomBoolean()) {
+            builder.field("bin", randomBytes, offset, length);
+        } else {
+            builder.field("bin").value(randomBytes, offset, length);
+        }
+        builder.endObject();
+
+        XContentParser parser = xcontentType().xContent().createParser(builder.bytes());
+        assertThat(parser.nextToken(), is(Token.START_OBJECT));
```

Comment:
`assertEquals` and `assertSame` for these? I like to avoid hamcrest if it doesn't buy anything beyond "normal" junit asserts.